django/db/migrations/operations/models.py

```diff
@@ -407,21 +407,38 @@ def database_forwards(self, app_label, schema_editor, from_state, to_state):
                     or not new_field.remote_field.through._meta.auto_created
                 ):
                     continue
-                # Rename the M2M table that's based on this model's name.
+
                 old_m2m_model = old_field.remote_field.through
                 new_m2m_model = new_field.remote_field.through
-                schema_editor.alter_db_table(
-                    new_m2m_model,
-                    old_m2m_model._meta.db_table,
-                    new_m2m_model._meta.db_table,
-                )
-                # Rename the column in the M2M table that's based on this
-                # model's name.
-                schema_editor.alter_field(
-                    new_m2m_model,
-                    old_m2m_model._meta.get_field(old_model._meta.model_name),
-                    new_m2m_model._meta.get_field(new_model._meta.model_name),
-                )
+
+                if (
+                    old_field.related_model._meta.model_name
+                    == old_model._meta.model_name
+                    or new_field.related_model._meta.model_name
+                    == new_model._meta.model_name
+                ):
+                    # Rename both columns and the table name of the M2M table
+                    # in case of same model names
+                    schema_editor._alter_many_to_many(
+                        new_m2m_model,
+                        old_field,
+                        new_field,
+                        strict=False,
+                    )
+                else:
+                    # Rename the M2M table that's based on this model's name.
+                    schema_editor.alter_db_table(
+                        new_m2m_model,
+                        old_m2m_model._meta.db_table,
+                        new_m2m_model._meta.db_table,
+                    )
+                    # Rename the column in the M2M table that's based on this
+                    # model's name.
+                    schema_editor.alter_field(
+                        new_m2m_model,
+                        old_m2m_model._meta.get_field(old_model._meta.model_name),
+                        new_m2m_model._meta.get_field(new_model._meta.model_name),
+                    )
```

Comment:
@DevilsAutumn I think there is a misunderstanding here. I think the point @felixxm and I were making is that instead of introducing changes to `alter_field` to allow the changes to be _forced_ `_alter_many_to_many` could be used directly instead, not that the whole `if`/`else` logic should be brought back

What you've done here is bring back the entire logic you had a few commits ago including the bug where you are passing `new_m2m_model` instead of `new_model` to `_alter_many_to_many`.

What is expected here is that you call `_alter_many_to_many` directly in all cases just like you did with `alter_field` previously while avoiding the method proxying and the associated requirement to introduce `alter_field(force)`.

Reply:
oh okay, i misunderstood. Just fixing